src/TParsec/Lexer.idr

```diff
@@ -0,0 +1,86 @@
+module TParsec.Lexer
+
+import Data.Tuple
+import Data.NEList
+import Data.Trie
+import TParsec.Position
+
+%default total
+%access public export
+
+record LexerParameters where
+  constructor MkLexerParameters
+-- Our lexer is parametrised over the type of tokens
+  Tok : Type
+-- We start with an association list between
+-- * keywords (as Strings)
+-- * keywords (as token values)
+  keywords : NEList (String, Tok)
+-- Some characters are special: they are separators, breaking a string into
+-- a list of tokens. Some are associated to a token value (e.g. parentheses)
+-- others are not (e.g. space)
+  breaking : Char -> (b ** if b then Maybe Tok else ())
+-- Finally, strings which are not decoded as keywords are coerced using a
+-- function to token values.
+  default  : String -> Tok
+
+LexResult : {p : LexerParameters} -> Type
+LexResult {p} = List (Position, Tok p)
+
```

Comment:
Given that we need to pass `p` explicitly everywhere,
we may as well make it an explicit argument.